tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd

```diff
@@ -0,0 +1,271 @@
+---
+title: Gaussian Process Latent Variable Model
+permalink: /:collection/:name/
+redirect_from: tutorials/12-gaussian-process-latent-variable-model/
+---
+
+# Gaussian Process Latent Variable Model
+
+In a previous tutorial, we have discussed latent variable models, in particular probabilistic principal
+component analysis (pPCA). Here, we show how we can extend the mapping provided by pPCA to non-linear mappings between input and output. For more details about the Gaussian Process Latent Variable Model (GPLVM), we refer the reader to the [original
+publication](https://jmlr.org/papers/v6/lawrence05a.html) and a [further extension](http://proceedings.mlr.press/v9/titsias10a/titsias10a.pdf).
+
+In short, the GPVLM is a dimensionality reduction technique that allows us to embed a high-dimensional dataset in a lower-dimensional embedding. Importantly, it provides the advantage that the linear mappings from the embedded space can be non-linearised through the use of Gaussian Processes.
+
+Let's start by loading some dependencies.
+```julia
+# Load Turing.
+using Turing
+using AbstractGPs, KernelFunctions, Random, Plots
+
+# Load other dependencies
+using Distributions, LinearAlgebra
+using VegaLite, DataFrames, StatsPlots, StatsBase
+using DelimitedFiles
+
+Random.seed!(1789);
+```
+
+We use a classical data set of synthetic data that is also used in the original publication.
+```julia
+oil_matrix = readdlm("Data.txt", Float64);
+labels = readdlm("DataLabels.txt", Float64);
+labels = mapslices(x -> findmax(x)[2], labels, dims=2);
+# normalize data
+dt = fit(ZScoreTransform, oil_matrix, dims=2);
+StatsBase.transform!(dt, oil_matrix);
+```
+
+We will start out, by demonstrating the basic similarity between pPCA (see the tutorial on this topic)
+and the GPLVM model. Indeed, pPCA is basically equivalent to running the GPLVM model with an
+automatic relevance determination (ARD) linear kernel.
+
+First, we re-introduce the pPCA model (see the tutorial on pPCA for details)
+```julia
+@model function pPCA(x)
+  # Dimensionality of the problem.
+  N, D = size(x)
+
+  # latent variable z
+  z ~ filldist(Normal(), D, N)
+
+  # weights/loadings W
+  w ~ filldist(Normal(0.0, 1.0), D, D)
+
+  # mean offset
+  mu = (w * z)'
+
+  for d in 1:D
+    x[:,d] ~ MvNormal(mu[:,d], 1.)
+  end
+end;
+```
+
+And here is the GPLVM model. The choice of kernel is determined by the kernel_function parameter.
+```julia
+@model function GPLVM(Y, kernel_function, ndim=4,::Type{T} = Float64) where {T}
+
+  # Dimensionality of the problem.
+  N, D = size(Y)
+  # dimensions of latent space
+  K = ndim
+  noise = 1e-3
+
+  # Priors
+  α ~ MvLogNormal(MvNormal(K, 1.0))
+  σ ~ LogNormal(0.0, 1.0)
+  Z ~ filldist(Normal(0., 1.), K, N)
+
+  kernel = kernel_function(α, σ)
+
+  ## DENSE GP
+  gp = GP(kernel)
+  prior = gp(ColVecs(Z), noise)
+
+  for d in 1:D
+    Y[:, d] ~ prior
```

Comment:
Spacing inconsistent. Here is a space (`Y[:, d]`) and above it is not (`x[:,d]`). There is no good reason to not stick to one style and use it everywhere. I would suggest using a space behind the comma. See also "Use whitespace to make the code more readable." at [BlueStyle](https://github.com/invenia/BlueStyle).